plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/models/OpenShiftProjectCache.java

```diff
@@ -53,20 +53,18 @@ public OpenShiftProjectCache() {
 	
 	@Override
 	public Collection<IProjectAdapter> getProjectsFor(final IOpenShiftConnection conn) {
-		String key = getCacheKey(conn);
-		Set<IProjectAdapter> adapters = getOrCreateEntry(key);
-		if (adapters.isEmpty()) {
+		Set<IProjectAdapter> adapters = getCacheFor(conn);
+		if(adapters == null) {
 			// new entry
 			Collection<IProject> projects = conn.<IProject>getResources(ResourceKind.PROJECT);
-			for (IProject project : projects) {
-				IProjectAdapter adapter = addNewProjectAdapter(adapters, conn, project);
-				notifyAdd(adapter);
+			adapters = createEntry(conn);
```

Comment:
here too (see below): what if 2 threads get in here concurrently, query the cache and create entries? checking if entry exists and creating an entry are not atomic, thus not thread safe I believe